src/stdlib_codata_type.fypp

```diff
@@ -0,0 +1,47 @@
+#:include "common.fypp"
+#:set KINDS = REAL_KINDS
+#:set PROPS = ("value", "uncertainty")
+module stdlib_codata_type
+    !! Codata constant type
+    use stdlib_kinds, only: #{for k in KINDS[:-1]}#${k}$, #{endfor}#${KINDS[-1]}$
+    use stdlib_io, only: FMT_REAL_DP
+    private
+
+    type, public :: codata_constant
+    !! Derived type for representing a Codata constant.
+        character(len=64) :: name
+        real(dp) :: value
+        real(dp) :: uncertainty
+        character(len=32) :: unit
+    contains 
+        procedure :: print
+        #:for p in PROPS
+        #:for k in KINDS
+        procedure :: eval_${p}$_${k}$
+        #:endfor
+        #:endfor
+        #:for p in PROPS
+        generic :: eval_${p}$ => #{for k in KINDS[:-1]}#eval_${p}$_${k}$, #{endfor}#eval_${p}$_${KINDS[-1]}$
+        #:endfor
+    end type
+    
+contains
+
+subroutine print(self)
+    !! Print out the constant's name, value, uncertainty and unit.
+    class(codata_constant), intent(in) :: self
+    print "(A64, SP, "//FMT_REAL_DP//" A5, "//FMT_REAL_DP//", X, A32)", self%name, self%value, "+/-", self%uncertainty, self%unit 
+end subroutine
+
+#:for p in PROPS
+#:for k in KINDS
+elemental real(${k}$) function eval_${p}$_${k}$(self, mold) result(r)
```

Comment:
It's great to have a mechanism that natively converts the constant DT to a `real` value. For consistency with `stdlib`, would it be nice to have a `to_real` function that lives both inside the derived type and as a separate interface?

https://github.com/fortran-lang/stdlib/blob/c3d847796638eb72283e2bb370ebf426a2a46b13/example/strings/example_to_string.f90#L19


https://github.com/fortran-lang/stdlib/blob/c3d847796638eb72283e2bb370ebf426a2a46b13/example/bitsets/example_bitsets_to_string.f90#L10

Reply:
Great idea. I'll implement it.

Reply:
The function `to_real` would be nice for the value and the uncertainty. It means that we would have 2 functions:
- `to_real_value`
- `to_real_uncertainty`
Or you meant that the function `to_real` would only apply to the value?

Reply:
I'll use an optional argument for getting the uncertainty and the default will be the value of the constant.

```fortran

#:for k in KINDS
elemental real(${k}$) function to_real_${k}$(self, mold, uncertainty) result(r)
    !! Evaluate the constant value or uncertainty ${p}$ for the kind ${k}$
    class(codata_constant), intent(in) :: self
    real(${k}$), intent(in) :: mold
    logical, intent(in), optional :: uncertainty
    logical :: u
    u = optval(uncertainty, .false.)
    if(u .eq. .true.)then
        r = real(self%uncertainty, kind(mold))
    else
        r = real(self%value, kind(mold))
    endif
end function
#:endfor
```
Identical mechanism for the function `to_real` that lives outside of the DT.

Reply:
LGTM